ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/worldstate/Pruner.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.worldstate;
+
+import tech.pegasys.pantheon.ethereum.chain.BlockAddedEvent;
+import tech.pegasys.pantheon.ethereum.chain.Blockchain;
+import tech.pegasys.pantheon.ethereum.core.BlockHeader;
+import tech.pegasys.pantheon.ethereum.core.Hash;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class Pruner {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final MarkSweepPruner pruningStrategy;
+  private final Blockchain blockchain;
+  private ExecutorService executorService;
+  private final long retentionPeriodInBlocks;
+  private final AtomicReference<State> state = new AtomicReference<>(State.IDLE);
+  private volatile long markBlockNumber = 0;
+  private volatile BlockHeader markedBlockHeader;
+  private long transientForkThreshold;
+
+  public Pruner(
+      final MarkSweepPruner pruningStrategy,
+      final Blockchain blockchain,
+      final ExecutorService executorService,
+      final long transientForkThreshold,
+      final long retentionPeriodInBlocks) {
+    this.pruningStrategy = pruningStrategy;
+    this.executorService = executorService;
+    this.blockchain = blockchain;
+    if (!(transientForkThreshold >= 0 && retentionPeriodInBlocks >= transientForkThreshold)) {
+      throw new IllegalArgumentException(
```

Comment:
Just a note that this exception might be a bit aggressive.  We can throw if the values are < 0,  otherwise if one value is just overshadowing the other, I'd log a warning.